subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/model/internal/IdeaDependenciesProvider.java

```diff
@@ -204,8 +207,28 @@ public void visitFileDependency(ResolvedArtifactResult artifact, boolean testDep
         }
 
         @Override
-        public void visitGradleApiDependency(ResolvedArtifactResult artifact, File sources, boolean testConfiguration) {
-            fileDependencies.add(new SingleEntryModuleLibrary(toPath(ideaModule, artifact.getFile()), null, toPath(ideaModule, sources), scope));
+        public void visitGradleApiDependency(ResolvedArtifactResult artifact, File sources, boolean testDependency) {
+            ModuleLibrary dependency = new ModuleLibrary(
+                Collections.singletonList(toPath(ideaModule, artifact.getFile())),
+                Collections.emptyList(),
+                collectGradleApiSources(sources),
+                Collections.emptySet(),
+                scope
+            );
+            fileDependencies.add(dependency);
+        }
+
+        private List<FilePath> collectGradleApiSources(File sources) {
+            if (sources == null) {
+                return Collections.emptyList();
+            }
+            File[] sourceDirectories = sources.listFiles(File::isDirectory);
+            if (sourceDirectories == null) {
+                return  Collections.emptyList();
+            }
+            return Collections.unmodifiableList(Arrays.stream(sourceDirectories)
+                .map(f -> toPath(ideaModule, f))
+                .collect(Collectors.toList()));
```

Comment:
I would like to use `Collectors.toUnmodifiableList()` here, but am unsure if it is fine to use Java10+ APIs here.

Reply:
Unfortunately, the minimum Java version for Gradle is 8.